### PR TITLE
Add support for Java super invocations

### LINKED ISSUE #99: Super calls not resolved
Example:
```java
public class A {
    public int getInt() {
        return 42;
    }
}

public class B extends A {
    @Override
    public int getInt() {
        return super.getInt() + 2;
    }
}
```

### PR DESCRIPTION
**Pull request by Masrepus**
Closes #99 and finishes a task in #37 (SuperExpr support)

### CODE REVIEW

File: src/main/java/de/fraunhofer/aisec/cpg/frontends/java/DeclarationHandler.java [hunk 1]
```diff
@@ -210,15 +209,22 @@ public RecordDeclaration handleClassOrInterfaceDeclaration(
...
     RecordDeclaration recordDeclaration =
-        NodeBuilder.newRecordDeclaration(name, superTypes, "class", classInterDecl.toString());
+        NodeBuilder.newRecordDeclaration(name, "class", classInterDecl.toString());
+    recordDeclaration.setSuperClasses(
+        classInterDecl.getExtendedTypes().stream()
+            .map(this.lang::getTypeAsGoodAsPossible)
+            .collect(Collectors.toList()));
+    if (recordDeclaration.getSuperClasses().isEmpty()) {
+      List<Type> superClasses = new ArrayList<>();
+      superClasses.add(TypeParser.createFrom("java.lang.Object", true));
```
Comment (JulianSchuette) [inline src/main/java/de/fraunhofer/aisec/cpg/frontends/java/DeclarationHandler.java:221]: Please do refer to class names as Strings but rather by `Object.class.getName()` to simplify refactoring.

File: src/main/java/de/fraunhofer/aisec/cpg/passes/VariableUsageResolver.java [hunk 2]
```diff
@@ -234,7 +248,17 @@ private void resolveFieldUsages(Node current, RecordDeclaration curClass) {
       Node base = memberExpression.getBase();
       Node member = memberExpression.getMember();
       if (base instanceof DeclaredReferenceExpression) {
-        base = resolveBase((DeclaredReferenceExpression) memberExpression.getBase());
+        if (lang instanceof JavaLanguageFrontend && base.getName().equals("super")) {
+          if (curClass != null && !curClass.getSuperClasses().isEmpty()) {
+            base = recordMap.get(curClass.getSuperClasses().get(0)).getThis();
+          } else {
+            // no explicit super type -> java.lang.Object
+            Type objectType = TypeParser.createFrom("java.lang.Object", true);
```
Comment (JulianSchuette) [inline src/main/java/de/fraunhofer/aisec/cpg/passes/VariableUsageResolver.java:256]: Same here - please use `Object.class.getName()`

**Comment by oxisto** [comment]
Now you broke it :D

File: src/test/java/de/fraunhofer/aisec/cpg/enhancements/SuperCallTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import de.fraunhofer.aisec.cpg.TestUtils;
+import de.fraunhofer.aisec.cpg.graph.*;
+import de.fraunhofer.aisec.cpg.helpers.Util;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.junit.jupiter.api.Test;
+
+public class SuperCallTest {
```
Comment (oxisto) [inline src/test/java/de/fraunhofer/aisec/cpg/enhancements/SuperCallTest.java:14]:
```suggestion
class SuperCallTest {
```

**Pull request merged**